src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <boost/function_output_iterator.hpp>
+
+// TODO: debug, remove
+#include <iostream>
+
+namespace osrm
+{
+namespace engine
+{
+namespace routing_algorithms
+{
+
+// Unqualified calls below are from the mld namespace.
+// This alternative implementation works only for mld.
+using namespace mld;
+
+using Heap = SearchEngineData<Algorithm>::QueryHeap;
+using Partition = partition::MultiLevelPartitionView;
+using Facade = datafacade::ContiguousInternalMemoryDataFacade<Algorithm>;
+
+// Implementation details
+namespace
+{
+
+// Alternative paths candidate via nodes are taken from overlapping search spaces.
+// Overlapping by a third guarantees us taking candidate nodes "from the middle".
+const constexpr auto kSearchSpaceOverlapFactor = 1.66;
+// Maximum number of alternative paths to return.
+const constexpr auto kMaxAlternatives = 3;
+// Maximum number of alternative paths to unpack and run high-quality checks on.
+// Unpack up to five times more pre-filtered paths than the user wants.
+// Note: this knob tunes how many unpacked paths go through high-quality filtering.
+// Set lower for higher performance but alternatives of lower quality.
+const constexpr auto kMaxAlternativesToUnpack = kMaxAlternatives * 5;
+// Alternative paths length requirement (stretch).
+// At most 25% longer then the shortest path.
+const constexpr auto kEpsilon = 0.25;
+// Alternative paths similarity requirement (sharing).
+// At least 15% different than the shortest path.
+const constexpr auto kGamma = 0.85;
+// Alternative paths are still reasonable around the via node candidate (local optimality).
+// At least optimal around 10% sub-paths around the via node candidate.
+const /*constexpr*/ auto kAlpha = 0.10;
+// Note: ^ ICEs gcc 7.1, just leave it out for better times
+
+// Represents a via middle node where forward (from s) and backward (from t)
+// search spaces overlap and the weight a path (made up of s,via and via,t) has.
+struct WeightedViaNode
+{
+    NodeID node;
+    EdgeWeight weight;
+};
+
+// Represents a complete packed path (made up of s,via and via,t)
+// its total weight and the via node used to construct the path.
+struct WeightedViaNodePackedPath
+{
+    WeightedViaNode via;
+    PackedPath path;
+};
+
+// Represents a high-detail unpacked path (s, .., via, .., t)
+// its total weight and the via node used to construct the path.
+struct WeightedViaNodeUnpackedPath
+{
+    WeightedViaNode via;
+    UnpackedNodes nodes;
+    UnpackedEdges edges;
+};
+
+// Filters candidates which are on not unique.
+// Returns an iterator to the uniquified range's new end.
+// Note: mutates the range in-place invalidating iterators.
+template <typename RandIt> RandIt filterViaCandidatesByUniqueNodeIds(RandIt first, RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
```

Comment:
The static asserts I use here and below are for erroring out early and with a nice line such as

    RandIt is not a random access iterator

or

    The value RandIt points to is not a type T

if you mix up the types here you will get a huge amount of garbage errors from within the stdlib. 

The sort call requires at least a `.nodes` property on the types. But writing traits for these things is a pita. This will be solved once we have Concepts (in C++3* maybe ..).